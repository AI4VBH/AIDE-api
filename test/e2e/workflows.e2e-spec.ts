import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { WorkflowsController } from 'modules/workflows/workflows.controller';
import { WorkflowsService } from 'modules/workflows/workflows.service';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import WorkflowMocks from '../test_data/mocks/workflows/workflows-index';
import { MonaiWorkflow } from 'modules/workflows/monai-workflow.interfaces';
import { RolesService } from 'modules/roles/roles.service';
import { KeycloakAdminService } from 'shared/keycloak/keycloak-admin.service';

const server = setupServer();
const testMonaiBasePath = 'https://localhost:7337';
const testMigBasePath = 'https://localhost:7338';

describe('/Workflows Integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          // we mock the path to the MONAI API
          load: [
            () => ({
              MONAI_API_HOST: testMonaiBasePath,
              MIG_API_HOST: testMigBasePath,
            }),
          ],
        }),
        // we register the HttpService with no stubbing! ðŸŽ‰
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [WorkflowsController],
      providers: [WorkflowsService, RolesService, KeycloakAdminService],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it.each([
    WorkflowMocks.basicWorkflows1,
    WorkflowMocks.basicWorkflows2,
    WorkflowMocks.basicWorkflows3,
    WorkflowMocks.basicWorkflows4,
  ])('(GET) /workflows with returned data', async (workflow) => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflows`,
        (request, response, context) => {
          return response(context.json(workflow));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/workflows?pageNumber=1&pageSize=10',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it('(GET) /workflows without returned data', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflows`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.emptyWorkflowData));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/workflows?pageNumber=1&pageSize=10',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /workflows correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflows`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/workflows?pageNumber=1&pageSize=10',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it.each([
    WorkflowMocks.singleWorkflow1,
    WorkflowMocks.singleWorkflow2,
    WorkflowMocks.singleWorkflow3,
  ])(
    '(GET) /workflows/:id with returned data',
    async (workflow: MonaiWorkflow) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflows/${workflow.workflow_id}`,
          (request, response, context) => {
            return response(context.json(workflow));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        `/workflows/${workflow.workflow_id}`,
      );
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(200);
    },
  );

  it('(GET) /workflows/:id with non-existent id', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca`,
        (request, response, context) => {
          return response(
            context.status(404),
            context.json(WorkflowMocks.nonExistentGetWorkflowError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(404);
  });

  it('(GET) /workflows/:id with invalid id', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflows/invalidID`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowMocks.invalidWorkflowIdError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/workflows/invalidID',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /workflows/:id correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it.each([
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.createdAETitle,
      201,
      WorkflowMocks.basicDestination1,
    ],
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.existsAETitle,
      409,
      WorkflowMocks.basicDestination1,
    ],
  ])(
    '(PUT) /workflows/:id success with returned data',
    async (putBody, aeResponse, aeStatus, destinationResponse) => {
      server.use(
        rest.put(
          `${testMonaiBasePath}/workflows/${putBody.workflow_id}`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.postPutResponse));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(context.status(aeStatus), context.json(aeResponse));
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(destinationResponse));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .put(`/workflows/${putBody.workflow_id}`)
        .send({ original_workflow_name: 'some body', workflow: putBody });
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(200);
    },
  );

  it.each([
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.invalidBodyAETitle,
      400,
      WorkflowMocks.basicDestination1,
    ],
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.existsAETitle,
      409,
      WorkflowMocks.basicDestination2,
    ],
  ])(
    '(PUT) /workflows/:id unsuccessful with ae or destination',
    async (putBody, aeResponse, aeStatus, destinationResponse) => {
      server.use(
        rest.put(
          `${testMonaiBasePath}/workflows/${putBody.workflow_id}`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.postPutResponse));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(context.status(aeStatus), context.json(aeResponse));
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(destinationResponse));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .put(`/workflows/${putBody.workflow_id}`)
        .send({ original_workflow_name: 'some body', workflow: putBody });
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(400);
    },
  );

  it('(PUT) /workflows/:id with non-existent id', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflows/0ea7b5b9-64ba-4841-b252-d6e312ef7e8d`,
        (request, response, context) => {
          return response(
            context.status(404),
            context.json(WorkflowMocks.nonExistentPutWorkflowError),
          );
        },
      ),
      rest.post(
        `${testMigBasePath}/config/ae`,
        (request, response, context) => {
          return response(
            context.status(201),
            context.json(WorkflowMocks.createdAETitle),
          );
        },
      ),
      rest.get(
        `${testMigBasePath}/config/destination`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.basicDestination1));
        },
      ),
    );
    const response = await await request(app.getHttpServer())
      .put('/workflows/0ea7b5b9-64ba-4841-b252-d6e312ef7e8d')
      .send({
        original_workflow_name: 'some body',
        workflow: WorkflowMocks.singleWorkflow1,
      });
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(404);
  });

  it('(PUT) /workflows/:id with invalid id', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflows/invalidID`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowMocks.invalidWorkflowIdError),
          );
        },
      ),
      rest.post(
        `${testMigBasePath}/config/ae`,
        (request, response, context) => {
          return response(
            context.status(201),
            context.json(WorkflowMocks.createdAETitle),
          );
        },
      ),
      rest.get(
        `${testMigBasePath}/config/destination`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.basicDestination1));
        },
      ),
    );
    const response = await await request(app.getHttpServer())
      .put('/workflows/invalidID')
      .send(WorkflowMocks.singleWorkflow1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it('(PUT) /workflows/:id with invalid body', async () => {
    server.use(
      rest.put(
        `${testMonaiBasePath}/workflows/0ea7b5b9-64ba-4841-b252-d6e312ef7e8d`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowMocks.invalidWorkflowBodyError),
          );
        },
      ),
      rest.post(
        `${testMigBasePath}/config/ae`,
        (request, response, context) => {
          return response(
            context.status(201),
            context.json(WorkflowMocks.createdAETitle),
          );
        },
      ),
      rest.get(
        `${testMigBasePath}/config/destination`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.basicDestination1));
        },
      ),
    );
    const response = await await request(app.getHttpServer())
      .put('/workflows/0ea7b5b9-64ba-4841-b252-d6e312ef7e8d')
      .send(WorkflowMocks.singleWorkflow1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(PUT) /workflows/:id correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.put(
          `${testMonaiBasePath}/workflows/${WorkflowMocks.singleWorkflow1.workflow_id}`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(
              context.status(201),
              context.json(WorkflowMocks.createdAETitle),
            );
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.basicDestination1));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .put(`/workflows/${WorkflowMocks.singleWorkflow1.workflow_id}`)
        .send(WorkflowMocks.singleWorkflow1);
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(DELETE) /workflows/:id with returned data', async () => {
    server.use(
      rest.delete(
        `${testMonaiBasePath}/workflows/${WorkflowMocks.singleWorkflow1.workflow_id}`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.singleWorkflow1));
        },
      ),
    );
    const response = await request(app.getHttpServer()).delete(
      `/workflows/${WorkflowMocks.singleWorkflow1.workflow_id}`,
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(200);
  });

  it('(DELETE) /workflows/:id with non-existent id', async () => {
    server.use(
      rest.delete(
        `${testMonaiBasePath}/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca`,
        (request, response, context) => {
          return response(
            context.status(404),
            context.json(WorkflowMocks.nonExistentGetWorkflowError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).delete(
      '/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(404);
  });

  it('(DELETE) /workflows/:id with invalid id', async () => {
    server.use(
      rest.delete(
        `${testMonaiBasePath}/workflows/invalidID`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowMocks.invalidWorkflowIdError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).delete(
      '/workflows/invalidID',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(DELETE) /workflows/:id correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.delete(
          `${testMonaiBasePath}/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).delete(
        '/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it.each([
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.createdAETitle,
      201,
      WorkflowMocks.basicDestination1,
    ],
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.existsAETitle,
      409,
      WorkflowMocks.basicDestination1,
    ],
  ])(
    '(POST) /workflows success with returned data',
    async (postBody, aeResponse, aeStatus, destinationResponse) => {
      server.use(
        rest.post(
          `${testMonaiBasePath}/workflows`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.postPutResponse));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(context.status(aeStatus), context.json(aeResponse));
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(destinationResponse));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .post(`/workflows`)
        .send(postBody);
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(201);
    },
  );

  it.each([
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.invalidBodyAETitle,
      400,
      WorkflowMocks.basicDestination1,
    ],
    [
      WorkflowMocks.singleWorkflow1,
      WorkflowMocks.existsAETitle,
      409,
      WorkflowMocks.basicDestination2,
    ],
  ])(
    '(POST) /workflows unsuccessful with ae or destination',
    async (postBody, aeResponse, aeStatus, destinationResponse) => {
      server.use(
        rest.post(
          `${testMonaiBasePath}/workflows`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.postPutResponse));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(context.status(aeStatus), context.json(aeResponse));
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(destinationResponse));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .post(`/workflows`)
        .send(postBody);
      expect(response.body).toMatchSnapshot();
      expect(response.statusCode).toBe(400);
    },
  );

  it('(POST) /workflows with invalid body', async () => {
    server.use(
      rest.post(
        `${testMigBasePath}/config/ae`,
        (request, response, context) => {
          return response(
            context.status(201),
            context.json(WorkflowMocks.createdAETitle),
          );
        },
      ),
      rest.get(
        `${testMigBasePath}/config/destination`,
        (request, response, context) => {
          return response(context.json(WorkflowMocks.basicDestination1));
        },
      ),
      rest.post(
        `${testMonaiBasePath}/workflows`,
        (request, response, context) => {
          return response(
            context.status(400),
            context.json(WorkflowMocks.invalidWorkflowBodyError),
          );
        },
      ),
    );
    const response = await await request(app.getHttpServer())
      .post('/workflows')
      .send(WorkflowMocks.singleWorkflow1);
    expect(response.body).toMatchSnapshot();
    expect(response.statusCode).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(POST) /workflows correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.post(
          `${testMonaiBasePath}/workflows`,
          (request, response, context) => {
            return response(context.status(code));
          },
        ),
        rest.post(
          `${testMigBasePath}/config/ae`,
          (request, response, context) => {
            return response(
              context.status(201),
              context.json(WorkflowMocks.createdAETitle),
            );
          },
        ),
        rest.get(
          `${testMigBasePath}/config/destination`,
          (request, response, context) => {
            return response(context.json(WorkflowMocks.basicDestination1));
          },
        ),
      );
      const response = await request(app.getHttpServer())
        .post('/workflows')
        .send(WorkflowMocks.singleWorkflow1);
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );
});
