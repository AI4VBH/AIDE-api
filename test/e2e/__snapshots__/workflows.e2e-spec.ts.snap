// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Workflows Integration Tests (DELETE) /workflows/:id with invalid id 1`] = `
{
  "detail": "Failed to validate id, not a valid guid",
  "instance": "/workflows/invalidID",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (DELETE) /workflows/:id with non-existent id 1`] = `
{
  "detail": "Failed to validate id, workflow not found",
  "instance": "/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/Workflows Integration Tests (DELETE) /workflows/:id with returned data 1`] = `
{
  "id": "f58fe206-1976-45bc-9d14-b9459eee2dfe",
  "revision": 3,
  "workflow": {
    "description": "This workflow is to be used for connectathon",
    "informatics_gateway": {
      "ae_title": "MonaiSCU",
      "data_origins": [
        "test",
      ],
      "export_destinations": [
        "ORTHANC",
      ],
    },
    "name": "Export Workflow",
    "tasks": [
      {
        "args": {
          "test": "test1",
        },
        "artifacts": {
          "input": [
            {
              "mandatory": true,
              "name": "study",
              "value": "{{ context.input.dicom }}",
            },
          ],
          "output": [],
        },
        "description": "Connectathon export task",
        "export_destinations": [
          {
            "name": "ORTHANC",
          },
        ],
        "id": "export-task-connectathon",
        "input_parameters": null,
        "ref": "",
        "task_destinations": [],
        "type": "export",
      },
    ],
    "version": "2.0.0",
  },
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (GET) /workflows with returned data 1`] = `
{
  "data": [
    {
      "description": "This workflow is to be used for connectathon",
      "id": "5fc10709-98cc-4d6a-8b68-06a722fa5dd7",
      "name": "Export Workflow",
      "revision": 1,
      "version": "1.0.0",
      "workflow_id": "ba85e76a-f03e-4b54-b2ea-c7736afa8c13",
    },
  ],
  "totalPages": 1,
  "totalRecords": 1,
}
`;

exports[`/Workflows Integration Tests (GET) /workflows with returned data 2`] = `
{
  "data": [
    {
      "description": ".",
      "id": "c7418628-e06b-49c1-9001-428cd05cec4e",
      "name": "Test workflow 1",
      "revision": 1,
      "version": "3",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
    {
      "description": "This workflow is to be used for connectathon",
      "id": "5fc10709-98cc-4d6a-8b68-06a722fa5dd7",
      "name": "Export Workflow",
      "revision": 1,
      "version": "1.0.0",
      "workflow_id": "ba85e76a-f03e-4b54-b2ea-c7736afa8c13",
    },
  ],
  "totalPages": 1,
  "totalRecords": 2,
}
`;

exports[`/Workflows Integration Tests (GET) /workflows with returned data 3`] = `
{
  "data": [
    {
      "description": ".",
      "id": "c7418628-e06b-49c1-9001-428cd05cec4e",
      "name": "Test workflow 1",
      "revision": 1,
      "version": "3",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
    {
      "description": "This workflow is to be used for connectathon",
      "id": "5fc10709-98cc-4d6a-8b68-06a722fa5dd7",
      "name": "Export Workflow",
      "revision": 1,
      "version": "1.0.0",
      "workflow_id": "ba85e76a-f03e-4b54-b2ea-c7736afa8c13",
    },
    {
      "description": "This workflow is to be used for connectathon",
      "id": "51725bf9-d7c6-49fa-bfd7-beb94a2a8733",
      "name": "Export Workflow",
      "revision": 2,
      "version": "1.0.0",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
  ],
  "totalPages": 1,
  "totalRecords": 3,
}
`;

exports[`/Workflows Integration Tests (GET) /workflows with returned data 4`] = `
{
  "data": [
    {
      "description": "This workflow is to be used for connectathon",
      "id": "f58fe206-1976-45bc-9d14-b9459eee2dfe",
      "name": "Export Workflow",
      "revision": 3,
      "version": "2.0.0",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
    {
      "description": ".",
      "id": "c7418628-e06b-49c1-9001-428cd05cec4e",
      "name": "Test workflow 1",
      "revision": 1,
      "version": "3",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
    {
      "description": "This workflow is to be used for connectathon",
      "id": "5fc10709-98cc-4d6a-8b68-06a722fa5dd7",
      "name": "Export Workflow",
      "revision": 1,
      "version": "1.0.0",
      "workflow_id": "ba85e76a-f03e-4b54-b2ea-c7736afa8c13",
    },
    {
      "description": "This workflow is to be used for connectathon",
      "id": "51725bf9-d7c6-49fa-bfd7-beb94a2a8733",
      "name": "Export Workflow",
      "revision": 2,
      "version": "1.0.0",
      "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
    },
    {
      "description": "Quite a long description, Lorem ipsum dolor sit amet, consectetur tincidunt.",
      "id": "12d15d1e-294d-4de6-856c-9d6c2d786ea6",
      "name": ".",
      "revision": 1,
      "version": "String",
      "workflow_id": "d5c9161e-92c6-412a-ab54-eb4f383fce56",
    },
  ],
  "totalPages": 1,
  "totalRecords": 5,
}
`;

exports[`/Workflows Integration Tests (GET) /workflows without returned data 1`] = `
{
  "data": [],
  "totalPages": 1,
  "totalRecords": 0,
}
`;

exports[`/Workflows Integration Tests (GET) /workflows/:id with invalid id 1`] = `
{
  "detail": "Failed to validate id, not a valid guid",
  "instance": "/workflows/invalidID",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (GET) /workflows/:id with non-existent id 1`] = `
{
  "detail": "Failed to validate id, workflow not found",
  "instance": "/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/Workflows Integration Tests (GET) /workflows/:id with returned data 1`] = `
{
  "id": "f58fe206-1976-45bc-9d14-b9459eee2dfe",
  "revision": 3,
  "workflow": {
    "description": "This workflow is to be used for connectathon",
    "informatics_gateway": {
      "ae_title": "MonaiSCU",
      "data_origins": [
        "test",
      ],
      "export_destinations": [
        "ORTHANC",
      ],
    },
    "name": "Export Workflow",
    "tasks": [
      {
        "args": {
          "test": "test1",
        },
        "artifacts": {
          "input": [
            {
              "mandatory": true,
              "name": "study",
              "value": "{{ context.input.dicom }}",
            },
          ],
          "output": [],
        },
        "description": "Connectathon export task",
        "export_destinations": [
          {
            "name": "ORTHANC",
          },
        ],
        "id": "export-task-connectathon",
        "input_parameters": null,
        "ref": "",
        "task_destinations": [],
        "type": "export",
      },
    ],
    "version": "2.0.0",
  },
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (GET) /workflows/:id with returned data 2`] = `
{
  "id": "5fc10709-98cc-4d6a-8b68-06a722fa5dd7",
  "revision": 1,
  "workflow": {
    "description": "This workflow is to be used for connectathon",
    "informatics_gateway": {
      "ae_title": "MonaiSCU",
      "data_origins": [
        "test",
      ],
      "export_destinations": [
        "ORTHANC",
      ],
    },
    "name": "Export Workflow",
    "tasks": [
      {
        "args": {
          "test": "test1",
        },
        "artifacts": {
          "input": [
            {
              "mandatory": true,
              "name": "study",
              "value": "{{ context.input.dicom }}",
            },
          ],
          "output": [],
        },
        "description": "Connectathon export task",
        "export_destinations": [
          {
            "name": "ORTHANC",
          },
        ],
        "id": "export-task-connectathon",
        "input_parameters": null,
        "ref": "",
        "task_destinations": [],
        "type": "export",
      },
    ],
    "version": "1.0.0",
  },
  "workflow_id": "ba85e76a-f03e-4b54-b2ea-c7736afa8c13",
}
`;

exports[`/Workflows Integration Tests (GET) /workflows/:id with returned data 3`] = `
{
  "id": "12d15d1e-294d-4de6-856c-9d6c2d786ea6",
  "revision": 1,
  "workflow": {
    "description": "Quite a long description, Lorem ipsum dolor sit amet, consectetur tincidunt.",
    "informatics_gateway": {
      "ae_title": "MonaiSCU",
      "data_origins": [
        "test",
      ],
      "export_destinations": [
        "ORTHANC",
      ],
    },
    "name": ".",
    "tasks": [
      {
        "args": {
          "test": "test1",
        },
        "artifacts": {
          "input": [
            {
              "mandatory": true,
              "name": "study",
              "value": "{{ context.input.dicom }}",
            },
          ],
          "output": [],
        },
        "description": "Connectathon export task",
        "export_destinations": [
          {
            "name": "ORTHANC",
          },
        ],
        "id": "export-task-connectathon",
        "input_parameters": null,
        "ref": "",
        "task_destinations": [],
        "type": "export",
      },
    ],
    "version": "String",
  },
  "workflow_id": "d5c9161e-92c6-412a-ab54-eb4f383fce56",
}
`;

exports[`/Workflows Integration Tests (POST) /workflows success with returned data 1`] = `
{
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (POST) /workflows success with returned data 2`] = `
{
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (POST) /workflows unsuccessful with ae or destination 1`] = `
{
  "detail": "'' is not a valid AE Title (source: MonaiApplicationEntity).",
  "status": 400,
  "title": "Validation error.",
  "traceId": "00-3795116b33338ae69581d64083acddc1-5cfe47801bc246c4-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (POST) /workflows unsuccessful with ae or destination 2`] = `
{
  "message": "Not all export destinations are registered",
  "statusCode": 400,
}
`;

exports[`/Workflows Integration Tests (POST) /workflows with invalid body 1`] = `
{
  "detail": "Failed to validate workflow: 'informaticsGateway' cannot be null (source: Artifact 1)., '' is not a valid AE Title (source: informaticsGateway)., '' is not a valid Informatics Gateway - exportDestinations (source: informaticsGateway)., Missing Workflow Version., Missing Workflow Tasks.",
  "instance": "/workflows",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-2d354c289398f59c2ad0247e5e8f667c-ea4c3161f31b70ea-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id success with returned data 1`] = `
{
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id success with returned data 2`] = `
{
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id unsuccessful with ae or destination 1`] = `
{
  "detail": "'' is not a valid AE Title (source: MonaiApplicationEntity).",
  "status": 400,
  "title": "Validation error.",
  "traceId": "00-3795116b33338ae69581d64083acddc1-5cfe47801bc246c4-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id unsuccessful with ae or destination 2`] = `
{
  "message": "Not all export destinations are registered",
  "statusCode": 400,
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id with invalid body 1`] = `
{
  "detail": "Failed to validate workflow: 'informaticsGateway' cannot be null (source: Artifact 1)., '' is not a valid AE Title (source: informaticsGateway)., '' is not a valid Informatics Gateway - exportDestinations (source: informaticsGateway)., Missing Workflow Version., Missing Workflow Tasks.",
  "instance": "/workflows",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-2d354c289398f59c2ad0247e5e8f667c-ea4c3161f31b70ea-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id with invalid id 1`] = `
{
  "detail": "Failed to validate id, not a valid guid",
  "instance": "/workflows/invalidID",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/Workflows Integration Tests (PUT) /workflows/:id with non-existent id 1`] = `
{
  "detail": "Failed to find workflow with Id: 0ea7b5b9-64ba-4841-b252-d6e312ef7e8d",
  "instance": "/workflows/0ea7b5b9-64ba-4841-b252-d6e312ef7e8d",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-c676afd543788d119ab730445e8e9b38-09a0ab1059e921e8-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;
